To review

Jean-BaptisteEmanuelZorg opened the pull request:
Adding new sortByField method

Jean-BaptisteEmanuelZorg committed:
rewrite sortByField method

Jean-BaptisteEmanuelZorg committed:
Item

Jean-BaptisteEmanuelZorg committed:
Start creating a method "compileBouquet"

Commit:
Correct bouquet proposal output

Jean-BaptisteEmanuelZorg committed:
Merge pull request #2 from Jean-BaptisteEmanuelZorg/Correct-bouquet-proposal

Correct bouquet proposal output

Jean-BaptisteEmanuelZorg committed:
"getColor" method problem

Jean-BaptisteEmanuelZorg committed:
the first version of the "compileBouquet" method

fox-rudie commented on src/main/java/FlowerStore/Item.java:
```diff
@@ -0,0 +1,38 @@
+package FlowerStore;
+
+public class Item {
+    private String itemName;
```
No need to duplicate the name of the entity on the field name.
Please, rename 'itemName' to 'name'

fox-rudie replied:
By the way, you can learn here how to use refactoring tools in IntelliJ idea. With the help of IDE, you can rename the field in the file and the usages.

fox-rudie commented on src/main/java/FlowerStore/ItemManager.java:
```diff
@@ -1,10 +1,23 @@
+        comparators.put("price", (i1, i2) -> i1.getPrice() - i2.getPrice());
+        comparators.put("description", (i1, i2) -> (i1.getDescription().compareTo(i2.getDescription())));
+
+        LIST_COMPARATOR = Collections.unmodifiableMap(comparators);
```
You can directly put them into the LIST_COMPARATORS map.

fox-rudie commented on src/main/java/FlowerStore/ItemManager.java:
```diff
@@ -18,48 +31,85 @@ public void removeItemById(Integer id) {
                 "items:\n" + items.toString();
     }
-
     public void sortByField(String field, SortOrder order){
```
Try to avoid creating of extra classes. This could be done with the use of "reverse" boolean. So, the could could be simplified:

```
public void sortByField(String field, boolean reverse) {
  Comparator<Item> comparator = LIST_COMPARATOR.get(field);
  Collections.sort(this.items, comparator);
  if (reverse) {
    Collections.reverse(this.items);
  }
}
```

fox-rudie replied:
This code doesn't return anything — what is the reason, then? 
Please, return the ordered items here.

fox-rudie commented on src/main/java/FlowerStore/Main.java:
```diff
@@ -1,20 +1,42 @@
+        List<BouquetFlower> bouquetProposal =  manager.compileBouquet(bouquetConfiguration,
+                manager.getItems());
+
+        System.out.println(bouquetProposal);
```
Here is going to be a new challenging task — write JUnit tests for this code. 
This is up to you how will write the test, but I'd like to have it covered with tests.
All methods from ItemsManager, by the way

fox-rudie commented on src/main/java/FlowerStore/ItemManager.java:
```diff
@@ -18,48 +31,85 @@ public void removeItemById(Integer id) {
+        String color = configuration.getColor();
+        int lowerBound = configuration.getPriceRange().getLowerBound();
+        int upperBound = configuration.getPriceRange().getUpperBound();
+        List<String> flower = configuration.getFlowers();
```
Should be plural because this doesn't make much sense then `String flowerType : flower`

fox-rudie commented on src/main/java/FlowerStore/ItemManager.java:
```diff
@@ -18,48 +31,85 @@ public void removeItemById(Integer id) {
+                        priceTotal -= bouquetFlower.getPrice();
+                        priceTotal += tempFlower.getPrice();
+                        baseFlowers.remove(tempFlower);
+                        System.out.println(priceTotal);
+                        System.out.println(bouquet);
```
Usually, we don't print with stdout. Use logger @Slf4j instead.

fox-rudie replied:
More about that: 
https://stackoverflow.com/questions/31869391/what-is-the-difference-between-java-logger-and-system-out-println

fox-rudie commented on src/main/java/FlowerStore/ItemManager.java:
```diff
@@ -18,48 +31,85 @@ public void removeItemById(Integer id) {
+
+        // logic
+
+        for (String flowerType : flower) {
```
Use streams here

fox-rudie commented on src/main/java/FlowerStore/ItemManager.java:
```diff
@@ -18,48 +31,85 @@ public void removeItemById(Integer id) {
+        for (String flowerType : flower) {
+            int i = 0;
+            for (BouquetFlower tempFlower : baseFlowers){
+                if (tempFlower.getItemName() == flowerType && tempFlower.getColor() == color && i < flowerTypeCount) {
```
It's almost always better to use equals instead of ==

fox-rudie replied:
In case of such complex booleans, it makes sense to introduce a variable that has reasonable name

fox-rudie commented on src/main/java/FlowerStore/ItemManager.java:
```diff
@@ -18,48 +31,85 @@ public void removeItemById(Integer id) {
-                    System.out.println("Sorted by price (DESC)");
+            }
+        }
+        Comparator<Item> comparator = LIST_COMPARATOR.get("price");
+        Collections.sort(bouquet, comparator);
+        Collections.reverse(bouquet);
+        Collections.sort(baseFlowers, comparator);
+        Collections.reverse(baseFlowers);
```
Comparator<Item> comparator = LIST_COMPARATOR.get("price");
        Collections.sort(bouquet, comparator.reversed());
        Collections.sort(baseFlowers, comparator.reversed());

fox-rudie commented on src/main/java/FlowerStore/ItemManager.java:
```diff
@@ -18,48 +31,85 @@ public void removeItemById(Integer id) {
+                .filter(item -> item instanceof BouquetFlower)
+                .map(item -> (BouquetFlower) item)
+                .collect(Collectors.toList());
+        List<BouquetFlower> bouquet= new ArrayList<>();
```
bouquetFlowers makes more sense to me

fox-rudie commented on src/main/java/FlowerStore/ItemManager.java:
```diff
@@ -18,48 +31,85 @@ public void removeItemById(Integer id) {
+
+        int i = 0;
+        while (priceTotal > upperBound){
+            for (BouquetFlower bouquetFlower : bouquet){
+                for (BouquetFlower tempFlower : baseFlowers){
```
It has complexity O(n^2) at least. I will try to think if it could be simplified

fox-rudie replied:
I am not sure, if it's possible to write it better but I will try to optimize this once I have time

fox-rudie commented:
Great work, overall!

Jean-BaptisteEmanuelZorg committed:
Refactored itemName to name.
Comparators are put into LIST_COMPARATOR directly.
Sort order changed to boolean.
sortByField method now have return statement.
ItemManager class covered with tests.
Logger added.
First block of logic from compileBouquet method rewritten using streams.
And other minor changes according to comments

Jean-BaptisteEmanuelZorg committed:
priceTotal atomic now.
Code cleaned and formatted
